feat(pipelines): refactor source stage and artifact handling

hoegertn opened the pull request:
refs #10872 
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: cd26951b71927704adf5adcfe70dfec6c56ea77d
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=29a70346-b131-424b-b834-062e88622de0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

hoegertn commented:
@rix0rrr as discussed the PR draft for the source action changes

What I did:
* Created an ISource interface to provide sources to CDK Pipelines
* Created implementations for all supported source providers
* Added autogeneration for sourceArtifact and cloudAssemblyArtifact
* Modified the SynthAction to accept the artifacts later in the process for auto-management

Feedback is much appreciated

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 535509f95f19e0ffa0b5b50be85715ce0c8b779d
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=951b8d02-39f0-403e-ba7f-53967a4f826d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

hoegertn commented:
The moment this is supported in CFN/CDK I would set this as default in the CodeCommitSource for CDK Pipelines.
https://aws.amazon.com/de/about-aws/whats-new/2020/11/aws-codepipeline-source-action-aws-codecommit-supports-git-clone/

This si what developers expect when accessing the source artifact.

rix0rrr commented:
I think it's worth me investing some effort into making the underlying `Action` classes mutable where it makes sense, to make it easier to implement the classes here.

Will start on that right now.

hoegertn commented on packages/@aws-cdk/pipelines/lib/pipeline.ts:
```diff
@@ -21,19 +23,28 @@ export interface CdkPipelineProps {
    *
    * @default - Required unless `codePipeline` is given
    */
-  readonly sourceAction?: codepipeline.IAction;
```
I was thinking of a backward-compatible way but I think this will make the code unreadable. The reason for dev preview is to experiment and break APIs when needed imho.

hoegertn commented on packages/@aws-cdk/pipelines/lib/pipeline.ts:
```diff
@@ -21,19 +23,28 @@ export interface CdkPipelineProps {
+   * @default - A source artifact is automatically created
    */
-  readonly cloudAssemblyArtifact: codepipeline.Artifact;
+  readonly sourceArtifact?: codepipeline.Artifact;
```
We need to if we want users to be able to provide a custom CodePipeline with existing source stage.

hoegertn commented on packages/@aws-cdk/pipelines/lib/sources/bitbucket-source.ts:
```diff
@@ -0,0 +1,55 @@
+    //
+  }
+
+  public provideSourceAction(sourceArtifact: codepipeline.Artifact): codepipeline.IAction {
```
Honestly, I am not sure if flexibility is what I want to achieve here. I want the user to provide a source not to do arbitrary stuff with the pipeline.

hoegertn commented on packages/@aws-cdk/pipelines/lib/sources/bitbucket-source.ts:
```diff
@@ -0,0 +1,55 @@
+  /**
+   * The BitBucket repo slug. (e.g. "aws/aws-cdk")
+   */
+  readonly slug: string;
```
I am not attached to the name. I just used the working from BB/GH

hoegertn commented on packages/@aws-cdk/pipelines/lib/sources/bitbucket-source.ts:
```diff
@@ -0,0 +1,55 @@
+   * @example 'arn:aws:codestar-connections:us-east-1:123456789012:connection/12345678-abcd-12ab-34cdef5678gh'
+   * @see https://docs.aws.amazon.com/codepipeline/latest/userguide/connections-create.html
+   */
+  readonly connectionArn: string;
```
That is the way BB is currently configured and GH will move to this too with GithubSourcev2

hoegertn commented on packages/@aws-cdk/pipelines/lib/sources/github-source.ts:
```diff
@@ -0,0 +1,80 @@
+  /**
+   * The GitHub repo slug. (e.g. "aws/aws-cdk")
+   */
+  readonly slug: string;
```
see above

hoegertn commented on packages/@aws-cdk/pipelines/lib/sources/github-source.ts:
```diff
@@ -0,0 +1,80 @@
+   *
+   * @see https://docs.aws.amazon.com/codepipeline/latest/userguide/GitHub-create-personal-token-CLI.html
+   *
+   * @default - use a SecretsManager secret called GitHub
```
Maybe there is a way to check if it exists?

hoegertn commented on packages/@aws-cdk/pipelines/lib/synths/simple-synth-action.ts:
```diff
@@ -186,10 +154,26 @@ export interface AdditionalArtifact {
+ * CodePipeline action to synthesize CDK applications.
+ * The source artifact and the cloud assembly artifact are lazily initialized by the CDK pipeline.
+ */
+export interface ISynthAction extends codepipeline.IAction {
```
Good catch. It definitely should not be an IAction

hoegertn commented:
> I think it's worth me investing some effort into making the underlying `Action` classes mutable where it makes sense, to make it easier to implement the classes here.
> 
> Will start on that right now.

What exactly do you want to make easier by this? Do you want to get rid of the factory style?

hoegertn commented on packages/@aws-cdk/pipelines/lib/synths/simple-synth-action.ts:
```diff
@@ -186,10 +154,26 @@ export interface AdditionalArtifact {
+   * @param sourceArtifact The source artifact of the CodePipeline
+   * @param cloudAssemblyArtifact The artifact where the CloudAssembly should be emitted
+   */
+  configureArtifacts(sourceArtifact: codepipeline.Artifact, cloudAssemblyArtifact: codepipeline.Artifact): void;
```
I am a big fan of concrete APIs that exactly tell the user what to do, but we can switch to a more generic approach.

hoegertn commented on packages/@aws-cdk/pipelines/lib/pipeline.ts:
```diff
@@ -21,19 +23,28 @@ export interface CdkPipelineProps {
    * @default - Required unless `codePipeline` or `sourceAction` is given
    */
-  readonly synthAction?: codepipeline.IAction;
+  readonly synthAction?: ISynthAction;
```
Sure, will handle this when we discussed the rest.

rix0rrr commented:
Superseded by https://github.com/aws/aws-cdk/pull/12326
